lib/internal/simulation_manager_mixins/characteristics_mixin.dart

```diff
@@ -40,150 +37,167 @@ mixin CharacteristicsMixin on SimulationManagerBaseWithErrorChecks {
-    await _errorIfCharacteristicNotReadable(targetCharacteristic);
-    Uint8List value = await targetCharacteristic.read();
-    await _errorIfDisconnected(peripheralId);
-    return CharacteristicResponse(targetCharacteristic, value);
-  }
-
-  Future<CharacteristicResponse> _readCharacteristicForService(
+    String transactionId,
+  ) =>
+      _saveCancelableOperation(transactionId, () async {
+        await _errorIfNotConnected(peripheralId);
+        SimulatedPeripheral targetPeripheral = _peripherals[peripheralId];
+        await _errorIfDiscoveryNotDone(targetPeripheral);
+
+        SimulatedCharacteristic targetCharacteristic = targetPeripheral
+            .getCharacteristicForService(serviceUuid, characteristicUUID);
+
+        await _errorIfCharacteristicIsNull(
+            targetCharacteristic, characteristicUUID);
+        await _errorIfCharacteristicNotReadable(targetCharacteristic);
+        Uint8List value = await targetCharacteristic.read();
+        await _errorIfDisconnected(peripheralId);
+        return CharacteristicResponse(targetCharacteristic, value);
+      });
+
+  Future<dynamic> _readCharacteristicForService(
     int serviceIdentifier,
     String characteristicUUID,
-  ) async {
-    SimulatedCharacteristic targetCharacteristic =
-        _findCharacteristicForServiceId(serviceIdentifier, characteristicUUID);
+    String transactionId,
+  ) =>
+      _saveCancelableOperation(transactionId, () async {
+        SimulatedCharacteristic targetCharacteristic =
+            _findCharacteristicForServiceId(
+                serviceIdentifier, characteristicUUID);
 
-    await _errorIfNotConnected(
-        _findPeripheralWithServiceId(serviceIdentifier).id);
+        await _errorIfNotConnected(
+            _findPeripheralWithServiceId(serviceIdentifier).id);
 
-    await _errorIfDiscoveryNotDone(
-        _findPeripheralWithServiceId(serviceIdentifier));
+        await _errorIfDiscoveryNotDone(
+            _findPeripheralWithServiceId(serviceIdentifier));
 
-    await _errorIfCharacteristicIsNull(
-        targetCharacteristic, characteristicUUID);
+        await _errorIfCharacteristicIsNull(
+            targetCharacteristic, characteristicUUID);
 
-    await _errorIfCharacteristicNotReadable(targetCharacteristic);
-    Uint8List value = await targetCharacteristic.read();
-    await _errorIfDisconnected(
-        _findPeripheralWithServiceId(serviceIdentifier).id);
-    return CharacteristicResponse(targetCharacteristic, value);
-  }
+        await _errorIfCharacteristicNotReadable(targetCharacteristic);
+        Uint8List value = await targetCharacteristic.read();
+        await _errorIfDisconnected(
+            _findPeripheralWithServiceId(serviceIdentifier).id);
+        return CharacteristicResponse(targetCharacteristic, value);
+      });
 
-  Future<SimulatedCharacteristic> _writeCharacteristicForIdentifier(
+  Future<dynamic> _writeCharacteristicForIdentifier(
```

Comment:
Either make saveCancelableOperation generic or cast its result to proper type; dynamic is a no-no.